lib/Sema/CSBindings.cpp

```diff
@@ -1215,8 +1215,12 @@ bool BindingSet::isViable(PotentialBinding &binding, bool isTransitive) {
     // as a binding because `$T0` could be inferred to
     // `(key: String, value: Int)` and binding `$T1` to `Array<(String, Int)>`
     // eagerly would be incorrect.
-    if (existing->Kind != binding.Kind)
-      continue;
+    if (existing->Kind != binding.Kind) {
+      // Array, Set and Dictionary allow conversions, everything else
+      // requires their generic arguments to match exactly.
+      if (existingType->isKnownStdlibCollectionType())
```

Comment:
What about Optional?

Reply:
Maybe this check should also call `hasConversions()`?

Reply:
We allow injections to optional but not conversions of its generic arguments, this only covers situations like `Int? conv $T` and `$T conv $U?`.

Reply:
This spot a lot narrower so I don't think we can use `hasConversions` here.

Reply:
> We allow injections into optional but not conversions of its generic arguments,

The optional-to-optional conversion is a lot like `Array` upcast isn't it?

Reply:
Somewhat but it wasn’t allowed previously but the logic here so we are not changing the behavior.

Reply:
Expressions like:

```
func test<T>(_: T, _: () -> T?) {}

class A {
}

class B : A {}

test(A()) {
    Optional(B())
}
```

Still work fine because optional conversion differ from collection upcast in that solver doesn't introduce a separate type variable when they are bound so even if we infer result type to be `B?` it would still much fine against `A?` from context, unlike what happens for collections where we'd infer the element and match it directly.

Reply:
Also allowing both `Int?` and `$T?` bindings results in the same solution multiple times, something that CSRanking doesn't properly handle today because we don't expect that to happen without score or binding differences, that's most likely the logic didn't have change check, I'd rather we revisit this whole method in the future than try to fix it now...